src/realm/object-store/c_api/app.cpp

```diff
@@ -809,6 +812,211 @@ RLM_API char* realm_user_get_refresh_token(const realm_user_t* user)
+    });
+}
+
+static void cb_proxy_for_completion(realm_userdata_t userdata, const realm_app_error_t* err)
+{
+    SyncUser::CompletionHandler* cxx_cb = static_cast<SyncUser::CompletionHandler*>(userdata);
+    REALM_ASSERT(cxx_cb);
+    std::optional<AppError> cxx_err;
+    if (err) {
+        std::optional<int> additional_error_code;
+        if (err->http_status_code) {
+            additional_error_code = err->http_status_code;
+        }
+        cxx_err =
+            AppError(ErrorCodes::Error(err->error), err->message, err->link_to_server_logs, additional_error_code);
+    }
+    (*cxx_cb)(cxx_err);
+    delete cxx_cb;
+}
+
+struct CAPIAppUser : SyncUser {
+    void* m_userdata = nullptr;
+    realm_free_userdata_func_t m_free = nullptr;
+    const std::string m_app_id;
+    const std::string m_user_id;
+    realm_user_get_access_token_cb_t m_access_token_cb = nullptr;
+    realm_user_get_refresh_token_cb_t m_refresh_token_cb = nullptr;
+    realm_user_state_cb_t m_state_cb = nullptr;
+    realm_user_access_token_refresh_required_cb_t m_atrr_cb = nullptr;
+    realm_user_get_sync_manager_cb_t m_sync_manager_cb = nullptr;
+    realm_user_request_log_out_cb_t m_request_log_out_cb = nullptr;
+    realm_user_request_refresh_user_cb_t m_request_refresh_user_cb = nullptr;
+    realm_user_request_refresh_location_cb_t m_request_refresh_location_cb = nullptr;
+    realm_user_request_access_token_cb_t m_request_access_token_cb = nullptr;
+    realm_user_track_realm_cb_t m_track_realm_cb = nullptr;
+    realm_user_create_file_action_cb_t m_create_fa_cb = nullptr;
+
+    CAPIAppUser(const char* app_id, const char* user_id)
+        : m_app_id(app_id)
+        , m_user_id(user_id)
+    {
+    }
+    CAPIAppUser(CAPIAppUser&& other)
+        : m_userdata(std::exchange(other.m_userdata, nullptr))
+        , m_free(std::exchange(other.m_free, nullptr))
+        , m_app_id(std::move(other.m_app_id))
+        , m_user_id(std::move(other.m_user_id))
+        , m_access_token_cb(std::move(other.m_access_token_cb))
+        , m_refresh_token_cb(std::move(other.m_refresh_token_cb))
+        , m_state_cb(std::move(other.m_state_cb))
+        , m_atrr_cb(std::move(other.m_atrr_cb))
+        , m_sync_manager_cb(std::move(other.m_sync_manager_cb))
+        , m_request_log_out_cb(std::move(other.m_request_log_out_cb))
+        , m_request_refresh_user_cb(std::move(other.m_request_refresh_user_cb))
+        , m_request_refresh_location_cb(std::move(other.m_request_refresh_location_cb))
+        , m_request_access_token_cb(std::move(other.m_request_access_token_cb))
+        , m_track_realm_cb(std::move(other.m_track_realm_cb))
+        , m_create_fa_cb(std::move(other.m_create_fa_cb))
+    {
+    }
+
+    ~CAPIAppUser()
+    {
+        if (m_free)
+            m_free(m_userdata);
+    }
+    std::string user_id() const noexcept override
+    {
+        return m_user_id;
+    }
+    std::string app_id() const noexcept override
```

Comment:
What is this being used for?

Reply:
It is used by the audit feature to distinguish realm identities.